File: src/domain/models/domain_model.py
"""DomainModel aggregate root for the Domain Model bounded context.

Manages the complete set of DDD artifacts for a project: domain stories,
ubiquitous language, bounded contexts, context relationships, and aggregate
designs. Enforces 4 invariants on finalize().

Invariants (from DDD.md §5):
1. Every UL term must appear in at least one DomainStory (word boundary match).
2. Every BoundedContext must have a SubdomainClassification.
3. Every Core subdomain must have at least one AggregateDesign.
4. Ambiguous terms must have per-context definitions.
"""

from __future__ import annotations

import re
import uuid
from typing import TYPE_CHECKING

from src.domain.models.domain_values import (
    AggregateDesign,
    BoundedContext,
    ContextRelationship,
    DomainStory,
    SubdomainClassification,
)
from src.domain.models.errors import DuplicateStoryError, InvariantViolationError
from src.domain.models.ubiquitous_language import UbiquitousLanguage

if TYPE_CHECKING:
    from src.domain.events.domain_events import DomainModelGenerated


class DomainModel:
    """Aggregate root: the complete DDD artifact set for a project.

    Attributes:
        model_id: Unique identifier for this domain model.
    """

    def __init__(self) -> None:
        self.model_id: str = str(uuid.uuid4())
        self._stories: list[DomainStory] = []
        self._language: UbiquitousLanguage = UbiquitousLanguage()
        self._contexts: list[BoundedContext] = []
        self._relationships: list[ContextRelationship] = []
        self._aggregates: list[AggregateDesign] = []
        self._events: list[DomainModelGenerated] = []

    # -- Properties -----------------------------------------------------------

    @property
    def domain_stories(self) -> tuple[DomainStory, ...]:
        """All domain stories (defensive copy)."""
        return tuple(self._stories)

    @property
    def ubiquitous_language(self) -> UbiquitousLanguage:
        """The ubiquitous language glossary."""
        return self._language

    @property
    def bounded_contexts(self) -> tuple[BoundedContext, ...]:
        """All bounded contexts (defensive copy)."""
        return tuple(self._contexts)

    @property
    def context_relationships(self) -> tuple[ContextRelationship, ...]:
        """All context relationships (defensive copy)."""
        return tuple(self._relationships)

    @property
    def aggregate_designs(self) -> tuple[AggregateDesign, ...]:
        """All aggregate designs (defensive copy)."""
        return tuple(self._aggregates)

    @property
    def events(self) -> tuple[DomainModelGenerated, ...]:
        """Domain events produced by this aggregate (defensive copy)."""
        return tuple(self._events)

    # -- Commands -------------------------------------------------------------

    def add_domain_story(self, story: DomainStory) -> None:
        """Add a business process narrative.

        Args:
            story: The domain story to add.

        Raises:
            DuplicateStoryError: If a story with the same name already exists.
            ValueError: If story name is empty.
        """
        if not story.name.strip():
            msg = "Story name cannot be empty"
            raise ValueError(msg)

        existing_names = {s.name.lower() for s in self._stories}
        if story.name.lower() in existing_names:
            msg = f"Domain story '{story.name}' already exists"
            raise DuplicateStoryError(msg)

        self._stories.append(story)

    def add_term(
        self,
        term: str,
        definition: str,
        context_name: str,
        source_question_ids: tuple[str, ...] = (),
    ) -> None:
        """Add a term to the ubiquitous language glossary.

        Args:
            term: The domain term.
            definition: What it means in this context.
            context_name: Which bounded context it belongs to.
            source_question_ids: Which questions surfaced this term.
        """
        self._language.add_term(term, definition, context_name, source_question_ids)

    def reassign_terms_to_context(
        self,
        from_context: str,
        to_context: str,
    ) -> None:
        """Move all terms from one context to another.

        Args:
            from_context: Source context name (case-insensitive match).
            to_context: Target context name.

        Raises:
            ValueError: If either context name is empty.
        """
        if not from_context.strip():
            msg = "Source context name cannot be empty"
            raise ValueError(msg)
        if not to_context.strip():
            msg = "Target context name cannot be empty"
            raise ValueError(msg)

        from src.domain.models.ubiquitous_language import TermEntry

        from_lower = from_context.lower()
        new_terms: list[TermEntry] = []
        for t in self._language._terms:
            if t.context_name.lower() == from_lower:
                new_terms.append(
                    TermEntry(
                        term=t.term,
                        definition=t.definition,
                        context_name=to_context,
                        source_question_ids=t.source_question_ids,
                    )
                )
            else:
                new_terms.append(t)
        self._language._terms = new_terms

    def add_bounded_context(self, context: BoundedContext) -> None:
        """Add a bounded context.

        Args:
            context: The bounded context to add.

        Raises:
            ValueError: If context name is empty or duplicate.
        """
        if not context.name.strip():
            msg = "Context name cannot be empty"
            raise ValueError(msg)

        existing_names = {c.name.lower() for c in self._contexts}
        if context.name.lower() in existing_names:
            msg = f"Bounded context '{context.name}' already exists"
            raise ValueError(msg)

        self._contexts.append(context)

    def classify_subdomain(
        self,
        context_name: str,
        classification: SubdomainClassification,
        rationale: str = "",
    ) -> None:
        """Classify a bounded context's subdomain type.

        Replaces the context entry with an updated classification.

        Args:
            context_name: Name of the bounded context to classify.
            classification: Core, Supporting, or Generic.
            rationale: Why this classification was chosen.

        Raises:
            ValueError: If the context does not exist.
        """
        for i, ctx in enumerate(self._contexts):
            if ctx.name.lower() == context_name.lower():
                self._contexts[i] = BoundedContext(
                    name=ctx.name,
                    responsibility=ctx.responsibility,
                    key_domain_objects=ctx.key_domain_objects,
                    classification=classification,
                    classification_rationale=rationale,
                )
                return

        msg = f"Bounded context '{context_name}' not found"
        raise ValueError(msg)

    def add_context_relationship(self, relationship: ContextRelationship) -> None:
        """Add a relationship between two bounded contexts.

        Args:
            relationship: The context relationship to add.

        Raises:
            ValueError: If upstream or downstream context name is empty.
        """
        if not relationship.upstream.strip() or not relationship.downstream.strip():
            msg = "Relationship upstream and downstream cannot be empty"
            raise ValueError(msg)
        self._relationships.append(relationship)

    def design_aggregate(self, aggregate: AggregateDesign) -> None:
        """Add an aggregate design for a bounded context.

        Args:
            aggregate: The aggregate design to add.

        Raises:
            ValueError: If aggregate name or context name is empty,
                or if a duplicate exists in the same context.
        """
        if not aggregate.name.strip():
            msg = "Aggregate name cannot be empty"
            raise ValueError(msg)
        if not aggregate.context_name.strip():
            msg = "Aggregate context name cannot be empty"
            raise ValueError(msg)

        existing = {(a.name.lower(), a.context_name.lower()) for a in self._aggregates}
        key = (aggregate.name.lower(), aggregate.context_name.lower())
        if key in existing:
            msg = (
                f"Aggregate design '{aggregate.name}' already exists"
                f" in context '{aggregate.context_name}'"
            )
            raise ValueError(msg)

        self._aggregates.append(aggregate)

    def finalize(self) -> None:
        """Validate all invariants and emit DomainModelGenerated.

        Raises:
            InvariantViolationError: If any invariant is violated.
        """
        self._check_terms_in_stories()
        self._check_context_classifications()
        self._check_core_aggregates()
        self._check_ambiguous_terms()

        from src.domain.events.domain_events import DomainModelGenerated

        self._events.append(
            DomainModelGenerated(
                model_id=self.model_id,
                domain_stories=tuple(self._stories),
                ubiquitous_language=self._language.terms,
                bounded_contexts=tuple(self._contexts),
                context_relationships=tuple(self._relationships),
                aggregate_designs=tuple(self._aggregates),
            )
        )

    # -- Invariant checks (private) -------------------------------------------

    def _check_terms_in_stories(self) -> None:
        """Invariant 1: Every UL term must appear in at least one DomainStory.

        Uses word boundary matching to prevent false positives from
        substring matches (e.g. "Or" should not match "Order").
        """
        story_text_parts: list[str] = []
        for story in self._stories:
            story_text_parts.append(story.name.lower())
            story_text_parts.extend(a.lower() for a in story.actors)
            story_text_parts.append(story.trigger.lower())
            story_text_parts.extend(s.lower() for s in story.steps)
            story_text_parts.extend(o.lower() for o in story.observations)
        story_text = " ".join(story_text_parts)

        for entry in self._language.terms:
            pattern = r"\b" + re.escape(entry.term.lower()) + r"\b"
            if not re.search(pattern, story_text):
                msg = f"Term '{entry.term}' not found in any domain story"
                raise InvariantViolationError(msg)

    def _check_context_classifications(self) -> None:
        """Invariant 2: Every BoundedContext must have a SubdomainClassification."""
        for ctx in self._contexts:
            if ctx.classification is None:
                msg = f"BoundedContext '{ctx.name}' has no classification"
                raise InvariantViolationError(msg)

    def _check_core_aggregates(self) -> None:
        """Invariant 3: Every Core subdomain must have at least one AggregateDesign."""
        core_contexts = {
            ctx.name
            for ctx in self._contexts
            if ctx.classification == SubdomainClassification.CORE
        }
        contexts_with_aggregates = {a.context_name for a in self._aggregates}

        for core_name in core_contexts:
            if core_name not in contexts_with_aggregates:
                msg = f"Core subdomain '{core_name}' has no aggregate design"
                raise InvariantViolationError(msg)

    def _check_ambiguous_terms(self) -> None:
        """Invariant 4: Ambiguous terms must have per-context definitions."""
        ambiguous = self._language.find_ambiguous_terms()
        for term in ambiguous:
            if not self._language.has_per_context_definitions(term):
                msg = f"Ambiguous term '{term}' needs per-context definitions"
                raise InvariantViolationError(msg)

File: internal/shared/domain/ddd/domain_model.go
package ddd

import (
	"fmt"
	"regexp"
	"strings"

	domainerrors "github.com/alty-cli/alty/internal/shared/domain/errors"
	"github.com/alty-cli/alty/internal/shared/domain/events"
	vo "github.com/alty-cli/alty/internal/shared/domain/valueobjects"
)

// DomainModel is the aggregate root for the Domain Model bounded context.
// It manages the complete set of DDD artifacts: domain stories, ubiquitous language,
// bounded contexts, context relationships, and aggregate designs.
//
// Invariants (checked on Finalize):
//  1. Every UL term must appear in at least one DomainStory (word boundary match).
//  2. Every BoundedContext must have a SubdomainClassification.
//  3. Every Core subdomain must have at least one AggregateDesign.
//  4. Ambiguous terms must have per-context definitions.
type DomainModel struct {
	modelID       string
	stories       []vo.DomainStory
	language      *UbiquitousLanguage
	contexts      []vo.DomainBoundedContext
	relationships []vo.ContextRelationship
	aggregates    []vo.AggregateDesign
	domainEvents  []events.DomainModelGenerated
	warnings      []string
}

// NewDomainModel creates a new DomainModel aggregate root with a generated ID.
func NewDomainModel(modelID string) *DomainModel {
	return &DomainModel{
		modelID:  modelID,
		language: NewUbiquitousLanguage(),
	}
}

// ModelID returns the model identifier.
func (dm *DomainModel) ModelID() string { return dm.modelID }

// DomainStories returns a defensive copy of all domain stories.
func (dm *DomainModel) DomainStories() []vo.DomainStory {
	out := make([]vo.DomainStory, len(dm.stories))
	copy(out, dm.stories)
	return out
}

// UbiquitousLanguage returns the ubiquitous language entity.
func (dm *DomainModel) UbiquitousLanguage() *UbiquitousLanguage {
	return dm.language
}

// BoundedContexts returns a defensive copy of all bounded contexts.
func (dm *DomainModel) BoundedContexts() []vo.DomainBoundedContext {
	out := make([]vo.DomainBoundedContext, len(dm.contexts))
	copy(out, dm.contexts)
	return out
}

// ContextRelationships returns a defensive copy of all context relationships.
func (dm *DomainModel) ContextRelationships() []vo.ContextRelationship {
	out := make([]vo.ContextRelationship, len(dm.relationships))
	copy(out, dm.relationships)
	return out
}

// AggregateDesigns returns a defensive copy of all aggregate designs.
func (dm *DomainModel) AggregateDesigns() []vo.AggregateDesign {
	out := make([]vo.AggregateDesign, len(dm.aggregates))
	copy(out, dm.aggregates)
	return out
}

// Events returns a defensive copy of domain events produced by this aggregate.
func (dm *DomainModel) Events() []events.DomainModelGenerated {
	out := make([]events.DomainModelGenerated, len(dm.domainEvents))
	copy(out, dm.domainEvents)
	return out
}

// Warnings returns a defensive copy of warnings produced during Finalize.
func (dm *DomainModel) Warnings() []string {
	out := make([]string, len(dm.warnings))
	copy(out, dm.warnings)
	return out
}

// -- Commands -----------------------------------------------------------------

// AddDomainStory adds a business process narrative.
func (dm *DomainModel) AddDomainStory(story vo.DomainStory) error {
	if strings.TrimSpace(story.Name()) == "" {
		return fmt.Errorf("story name cannot be empty")
	}

	nameLower := strings.ToLower(story.Name())
	for _, s := range dm.stories {
		if strings.ToLower(s.Name()) == nameLower {
			return fmt.Errorf("domain story '%s' already exists: %w",
				story.Name(), domainerrors.ErrAlreadyExists)
		}
	}

	dm.stories = append(dm.stories, story)
	return nil
}

// AddTerm adds a term to the ubiquitous language glossary.
func (dm *DomainModel) AddTerm(term, definition, contextName string, sourceQuestionIDs []string) error {
	return dm.language.AddTerm(term, definition, contextName, sourceQuestionIDs)
}

// ReassignTermsToContext moves all terms from one context to another.
func (dm *DomainModel) ReassignTermsToContext(fromContext, toContext string) error {
	if strings.TrimSpace(fromContext) == "" {
		return fmt.Errorf("source context name cannot be empty")
	}
	if strings.TrimSpace(toContext) == "" {
		return fmt.Errorf("target context name cannot be empty")
	}

	fromLower := strings.ToLower(fromContext)
	oldTerms := dm.language.Terms()
	newTerms := make([]vo.TermEntry, 0, len(oldTerms))
	for _, t := range oldTerms {
		if strings.ToLower(t.ContextName()) == fromLower {
			newTerms = append(newTerms, t.WithContextName(toContext))
		} else {
			newTerms = append(newTerms, t)
		}
	}
	dm.language.SetTerms(newTerms)
	return nil
}

// AddBoundedContext adds a bounded context.
func (dm *DomainModel) AddBoundedContext(ctx vo.DomainBoundedContext) error {
	if strings.TrimSpace(ctx.Name()) == "" {
		return fmt.Errorf("context name cannot be empty")
	}

	nameLower := strings.ToLower(ctx.Name())
	for _, c := range dm.contexts {
		if strings.ToLower(c.Name()) == nameLower {
			return fmt.Errorf("bounded context '%s' already exists", ctx.Name())
		}
	}

	dm.contexts = append(dm.contexts, ctx)
	return nil
}

// ClassifySubdomain classifies a bounded context's subdomain type.
func (dm *DomainModel) ClassifySubdomain(contextName string, classification vo.SubdomainClassification, rationale string) error {
	nameLower := strings.ToLower(contextName)
	for i, ctx := range dm.contexts {
		if strings.ToLower(ctx.Name()) == nameLower {
			dm.contexts[i] = vo.NewDomainBoundedContext(
				ctx.Name(),
				ctx.Responsibility(),
				ctx.KeyDomainObjects(),
				&classification,
				rationale,
			)
			return nil
		}
	}
	return fmt.Errorf("bounded context '%s' not found", contextName)
}

// AddContextRelationship adds a relationship between two bounded contexts.
func (dm *DomainModel) AddContextRelationship(rel vo.ContextRelationship) error {
	if strings.TrimSpace(rel.Upstream()) == "" || strings.TrimSpace(rel.Downstream()) == "" {
		return fmt.Errorf("relationship upstream and downstream cannot be empty")
	}
	dm.relationships = append(dm.relationships, rel)
	return nil
}

// DesignAggregate adds an aggregate design for a bounded context.
func (dm *DomainModel) DesignAggregate(agg vo.AggregateDesign) error {
	if strings.TrimSpace(agg.Name()) == "" {
		return fmt.Errorf("aggregate name cannot be empty")
	}
	if strings.TrimSpace(agg.ContextName()) == "" {
		return fmt.Errorf("aggregate context name cannot be empty")
	}

	aggKey := strings.ToLower(agg.Name()) + "\x00" + strings.ToLower(agg.ContextName())
	for _, a := range dm.aggregates {
		existingKey := strings.ToLower(a.Name()) + "\x00" + strings.ToLower(a.ContextName())
		if existingKey == aggKey {
			return fmt.Errorf("aggregate design '%s' already exists in context '%s'",
				agg.Name(), agg.ContextName())
		}
	}

	dm.aggregates = append(dm.aggregates, agg)
	return nil
}

// Finalize validates all invariants and emits DomainModelGenerated.
func (dm *DomainModel) Finalize() error {
	dm.warnings = nil

	if err := dm.checkTermsInStories(); err != nil {
		return err
	}
	if err := dm.checkContextClassifications(); err != nil {
		return err
	}
	if err := dm.checkCoreAggregates(); err != nil {
		return err
	}
	if err := dm.checkAmbiguousTerms(); err != nil {
		return err
	}

	dm.domainEvents = append(dm.domainEvents, events.NewDomainModelGenerated(
		dm.modelID,
		dm.stories,
		dm.language.Terms(),
		dm.contexts,
		dm.relationships,
		dm.aggregates,
	))
	return nil
}

// -- Invariant checks (private) -----------------------------------------------

func (dm *DomainModel) checkTermsInStories() error {
	var parts []string
	for _, story := range dm.stories {
		parts = append(parts, strings.ToLower(story.Name()))
		for _, a := range story.Actors() {
			parts = append(parts, strings.ToLower(a))
		}
		parts = append(parts, strings.ToLower(story.Trigger()))
		for _, s := range story.Steps() {
			parts = append(parts, strings.ToLower(s))
		}
		for _, o := range story.Observations() {
			parts = append(parts, strings.ToLower(o))
		}
	}
	storyText := strings.Join(parts, " ")

	for _, entry := range dm.language.Terms() {
		pattern := `\b` + regexp.QuoteMeta(strings.ToLower(entry.Term())) + `\b`
		matched, err := regexp.MatchString(pattern, storyText)
		if err != nil || !matched {
			return fmt.Errorf("term '%s' not found in any domain story: %w",
				entry.Term(), domainerrors.ErrInvariantViolation)
		}
	}
	return nil
}

func (dm *DomainModel) checkContextClassifications() error {
	for _, ctx := range dm.contexts {
		if ctx.Classification() == nil {
			return fmt.Errorf("bounded context '%s' has no classification: %w",
				ctx.Name(), domainerrors.ErrInvariantViolation)
		}
	}

	if len(dm.contexts) > 0 {
		allGeneric := true
		for _, ctx := range dm.contexts {
			if *ctx.Classification() != vo.SubdomainGeneric {
				allGeneric = false
				break
			}
		}
		if allGeneric {
			dm.warnings = append(dm.warnings,
				"All bounded contexts are classified as Generic. "+
					"A project with no Core or Supporting subdomain likely "+
					"has misclassified contexts.")
		}
	}
	return nil
}

func (dm *DomainModel) checkCoreAggregates() error {
	coreContexts := make(map[string]struct{})
	for _, ctx := range dm.contexts {
		if ctx.Classification() != nil && *ctx.Classification() == vo.SubdomainCore {
			coreContexts[ctx.Name()] = struct{}{}
		}
	}

	contextsWithAggs := make(map[string]struct{})
	for _, a := range dm.aggregates {
		contextsWithAggs[a.ContextName()] = struct{}{}
	}

	for coreName := range coreContexts {
		if _, ok := contextsWithAggs[coreName]; !ok {
			return fmt.Errorf("core subdomain '%s' has no aggregate design: %w",
				coreName, domainerrors.ErrInvariantViolation)
		}
	}
	return nil
}

func (dm *DomainModel) checkAmbiguousTerms() error {
	ambiguous := dm.language.FindAmbiguousTerms()
	for _, term := range ambiguous {
		if !dm.language.HasPerContextDefinitions(term) {
			return fmt.Errorf("ambiguous term '%s' needs per-context definitions: %w",
				term, domainerrors.ErrInvariantViolation)
		}
	}
	return nil
}
